docs/src/main/asciidoc/qute-reference.adoc

```diff
@@ -0,0 +1,698 @@
+<3> Inject the configured `Engine` instance.
+
+=== Injecting Beans Directly In Templates
+
+A CDI bean annotated with `@Named` can be referenced in any template through the `inject` namespace:
+
+[source,html]
+----
+{inject:foo.price} <1>
+----
+<1> First, a bean with name `foo` is found and then used as the base object.
+
+All expressions using the `inject` namespace are validated during build.
+For the expression `inject:foo.price` the implementation class of the injected bean must either have the `price` property (e.g. a `getPrice()` method) or a matching <<template_extension_methods,template extension method>> must exist. 
+
+NOTE: A `ValueResolver` is also generated for all beans annotated with `@Named` so that it's possible to access its properties without reflection.
+
+=== Parameter Declarations
+
+It is possible to specify optional parameter declarations in a template.
+Quarkus attempts to validate all expressions that reference such parameters.
+If an invalid/incorrect expression is found the build fails.
+
+NOTE: Only properties are currently validated in expressions; virtual methods are currently ignored.
+
+[source,html]
+----
+{@org.acme.Foo foo} <1>
+<!DOCTYPE html>
+<html>
+<head>
+<meta charset="UTF-8">
+<title>Qute Hello</title>
+</head>
+<body>
+    <h1>{title}</h1> <2>
+    Hello {foo.message}! <3>
+</body>
+</html>
+----
+<1> Parameter declaration - maps `foo` to `org.acme.Foo`.
+<2> Not validated - not matching a param declaration.
+<3> This expression is validated. `org.acme.Foo` must have a property `message` or a matching template extension method must exist. 
+
+NOTE: A value resolver is also generated for all types used in parameter declarations so that it's possible to access its properties without reflection.
+
+==== Overriding Parameter Declarations
+
+[source,html]
+----
+{@org.acme.Foo foo}
+<!DOCTYPE html>
+<html>
+<head>
+<meta charset="UTF-8">
+<title>Qute Hello</title>
+</head>
+<body>
+    <h1>{foo.message}</h1> <1>
+    {#for foo in baz.foos}
+     <p>Hello {foo.message}!</p> <2>
+    {/for}
+</body>
+</html>
+----
+<1> Validated against `org.acme.Foo`.
+<2> Not validated - `foo` is overriden in the loop section.
+
+[[template_extension_methods]]
+=== Template Extension Methods
+
+A value resolver is automatically generated for a template extension method annotated with `@TemplateExtension`.
+The method must be static, must not return `void` and must accept at least one parameter. 
+The class of the first parameter is used to match the base object and the method name is used to match the property name.
+
+[source,java]
+----
+package org.acme;
+
+class Item {
+
+    public final BigDecimal price;
+
+    public Item(BigDecimal price) {
+        this.price = price;
+    }
+}
+
+class MyExtensions {
+
+  @TemplateExtension
+  static BigDecimal discountedPrice(Item item) { <1>
+     return item.getPrice().multiply(new BigDecimal("0.9"));
+  }
```

Comment:
This is minor but indentation is not consistent here. You have four spaces then 2.